lib/mongo/socket.rb

```diff
@@ -169,11 +169,28 @@ def eof?
     private
 
     def read_from_socket(length)
-      data = String.new
+      data = ('x'*length).force_encoding('BINARY')
       deadline = (Time.now + timeout) if timeout
+      # We want to have a fixed and reasonably small size buffer for reads
+      # because, for example, OpenSSL reads in 16 kb chunks max.
+      # Having a 16 mb buffer means there will be 1000 reads each allocating
+      # 16 mb of memory and using 16 kb of it.
+      #
+      # The binary encoding is important, otherwise ruby performs encoding
+      # conversions of some sort during the write into the buffer which
+      # kills performance
+      buf_size = read_buffer_size
+      buf = ('x'*buf_size).force_encoding('BINARY')
```

Comment:
Ditto here. Also, if we're doing this in multiple places, I think it might be useful to have a helper method called something like `binary_string_with_capacity(length)` and just call that everywhere.